stdlib/seq.mli

```diff
@@ -13,97 +13,807 @@
+    @since 4.14 *)
+
+val drop_while : ('a -> bool) -> 'a t -> 'a t
+(** [drop_while p xs] is the sequence [xs], deprived of the prefix
+    [take_while p xs].
+
+    @since 4.14 *)
+
+val uniq : ('a -> 'a -> bool) -> 'a t -> 'a t
+(** Provided the function [eq] defines an equality on elements,
+    [uniq eq xs] is the sequence [xs],
+    deprived of its adjacent duplicate elements.
+
+    Thus, [uniq eq xs] has no adjacent duplicate elements.
+    Furthermore, if [xs] is sorted according to an ordering
+    that is compatible with [eq],
+    then [uniq eq xs] has no duplicate elements at all.
+
+    [uniq eq xs] is equivalent to [map head (group eq xs)].
+
+    @since 4.14 *)
+
+val group : ('a -> 'a -> bool) -> 'a t -> 'a t t
+(** Provided the function [eq] defines an equality on elements,
+    [group eq xs] is the sequence of the maximal runs
+    of adjacent duplicate elements of the sequence [xs].
+
+    Every element of [group eq xs] is a nonempty sequence of equal elements.
+
+    The concatenation [concat (group eq xs)] is equal to [xs].
+
+    Consuming [group eq xs], and consuming the sequences that it contains,
+    can cause [xs] to be consumed more than once. Therefore, [xs] must be
+    persistent.
+
+    @since 4.14 *)
+
+val memoize : 'a t -> 'a t
+(** The sequence [memoize xs] has the same elements as the sequence [xs].
+
+    Regardless of whether [xs] is ephemeral or persistent,
+    [memoize xs] is persistent: even if it is queried several times,
+    [xs] is queried at most once.
+    @since 4.14 *)
+
+exception ForcedTwice
+(** This exception is raised when a sequence returned by [once]
+    (or a suffix of it) is queried more than once.
+    @since 4.14 *)
+
+val once: 'a t -> 'a t
+(** The sequence [once xs] has the same elements as the sequence [xs].
+
+    Regardless of whether [xs] is ephemeral or persistent,
+    [once xs] is an ephemeral sequence: it can be queried at most once.
+    If it (or a suffix of it) is queried more than once, then the exception
+    [ForcedTwice] is raised. This can be useful, while debugging or testing,
+    to ensure that a sequence is consumed at most once.
+
+    @raise ForcedTwice if [once xs], or a suffix of it,
+           is queried more than once.
+    @since 4.14 *)
+
+val transpose : 'a t t -> 'a t t
+(** If [xss] is a matrix (a sequence of rows), then [transpose xss] is
+    the sequence of the columns of the matrix [xss].
+
+    The rows of the matrix [xss] are not required to have the same length.
+
+    The matrix [xss] is not required to be finite (in either direction).
+
+    The matrix [xss] must be persistent.
+
+    @since 4.14 *)
+
+val diagonals : 'a t t -> 'a t t
+(** If [xss] is a matrix (a sequence of rows), then [diagonals xss] is
+    the sequence of its diagonals.
+
+    The first diagonal contains just the first element of the
+    first row. The second diagonal contains the first element of the
+    second row and the second element of the first row; and so on.
```

Comment:
I'm not sure this diagnonals function is really useful anyway....

Reply:
The reason why `diagonals` is there is that it is used internally in the definition of `product`. But we could certainly remove it from the public interface if it seems too baroque.

Reply:
I am mildly in favor of having the function ("accept", but not "strong accept"), because I believe that this sort of functions is very useful when we want to control the behavior of our enumerations in presence of infinite sequences -- see my comment on zip/products.

I think that the documentation would benefit from some ascii art, explaning that, say

```
[a b c]
[d e]
[f g h]
```

becomes

```
[a]
[d b]
[f e c]
[g]
[h]
```

Finally: unlike @dbuenzli I wouldn't think of matrices here, but maybe we could gain his favorable opinion by calling the function `antidiagonals` instead of `diagonals`? <https://resources.wolframcloud.com/FunctionRepository/resources/Antidiagonal/> gives a crisp definition of what an antidiagonal is, "a list of elements for which the sum of the indices are equal", which might even suggest an alternative remark to document this function.

Reply:
I thought about matrices because the doc string told me so :-) 

That being said I share a bit the doubts of @xavierleroy about its usefulness. 

When I read the description I thought "ah that's a clever way of enumerating all the elements when you have a sequence of possibly infinite sequences". It certainly does so but is it useful to get these as a sequence of finite sequences ? Rather than simply a sequence of `'a` values (I know it's a `concat` away but if we concat immediately we could have a better name instead).

Reply:
I think the function that @dbuenzli describes is likely more immediately useful as a matrix enumeration (and corresponds with [Cantor's pairing function](https://en.wikipedia.org/wiki/Pairing_function)).

For this function, @gasche's suggestion of `antidiagonals` seems more expressive to me.